dll/win32/syssetup/wizard.c

```diff
@@ -410,69 +410,140 @@ DoWriteProductOption(PRODUCT_OPTION nOption)
+{
+    static const WCHAR s_szRosVersion[] = L"SYSTEM\\CurrentControlSet\\Control\\ReactOS\\Settings\\Version";
+    HKEY hKey;
+    LONG error;
+    DWORD cbData, dwValue;
+
     error = RegOpenKeyExW(HKEY_LOCAL_MACHINE, s_szRosVersion, 0, KEY_WRITE, &hKey);
     if (error)
         return FALSE;
 
     /* write ReportAsWorkstation value */
-    dwValue = (nOption == PRODUCT_OPTION_WORKSTATION);
+    switch (nOption)
+    {
+        case PRODUCT_OPTION_SERVER:
+            dwValue = 0;
+            break;
+        case PRODUCT_OPTION_WORKSTATION:
+            dwValue = 1;
+            break;
+    }
     cbData = sizeof(dwValue);
-    error = RegSetValueExW(hKey, L"ReportAsWorkstation", 0, REG_DWORD, (BYTE *)&dwValue, cbData);
+    error = RegSetValueExW(hKey, L"ReportAsWorkstation", 0, REG_DWORD, (const BYTE *)&dwValue, cbData);
 
     RegCloseKey(hKey);
-
     return error == ERROR_SUCCESS;
 }
 
-static void
-OnChooseServer(HWND hwndDlg)
+static BOOL
+WriteCSDVersion(PRODUCT_OPTION nOption)
 {
-    WCHAR szText[256];
+    static const WCHAR s_szWindowsNT[] = L"SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion";
+    static const WCHAR s_szControlWindows[] = L"SYSTEM\\CurrentControlSet\\Control\\Windows";
+    static const WCHAR s_szServicePack2[] = L"Service Pack 2";
+    static const WCHAR s_szServicePack3[] = L"Service Pack 3";
+    HKEY hKey1, hKey2;
+    LONG error1, error2;
+    DWORD cbData;
+
+    error1 = RegOpenKeyExW(HKEY_LOCAL_MACHINE, s_szWindowsNT, 0, KEY_WRITE, &hKey1);
+    error2 = RegOpenKeyExW(HKEY_LOCAL_MACHINE, s_szControlWindows, 0, KEY_WRITE, &hKey2);
+    if (error1 || error2)
+    {
+        if (hKey1)
+            RegCloseKey(hKey1);
+        if (hKey2)
+            RegCloseKey(hKey2);
+        return FALSE;
+    }
 
-    SetDlgItemTextW(hwndDlg, IDC_PRODUCT_SUITE, L"Terminal Server");
-    SetDlgItemTextW(hwndDlg, IDC_PRODUCT_TYPE, L"ServerNT");
+    /* write CSDVersion value */
+    switch (nOption)
+    {
+        case PRODUCT_OPTION_SERVER:
+            cbData = sizeof(s_szServicePack2);
+            error1 = RegSetValueExW(hKey1, L"CSDVersion", 0, REG_SZ, (const BYTE *)s_szServicePack2, cbData);
+            error2 = RegSetValueExW(hKey2, L"CSDVersion", 0, REG_SZ, (const BYTE *)s_szServicePack2, cbData);
+            break;
 
-    LoadStringW(hDllInstance, IDS_PRODUCTSERVERINFO, szText, _countof(szText));
-    SetDlgItemTextW(hwndDlg, IDC_PRODUCT_DESCRIPTION, szText);
+        case PRODUCT_OPTION_WORKSTATION:
+            cbData = sizeof(s_szServicePack3);
+            error1 = RegSetValueExW(hKey1, L"CSDVersion", 0, REG_SZ, (const BYTE *)s_szServicePack3, cbData);
+            error2 = RegSetValueExW(hKey2, L"CSDVersion", 0, REG_SZ, (const BYTE *)s_szServicePack3, cbData);
+            break;
+    }
+
+    RegCloseKey(hKey1);
+    RegCloseKey(hKey2);
+
+    return !error1 && !error2;
+}
+
+static BOOL
+DoWriteProductOption(PRODUCT_OPTION nOption)
+{
+    return WriteProductSuiteAndType(nOption) &&
```

Comment:
Why the separation? I don't get it - isn't it simpler to just write all registry values in a single `case` or `if` statement?

Reply:
United.